Data/ByteString/Lazy/ReadInt.hs

```diff
@@ -0,0 +1,229 @@
+readInt :: ByteString -> Maybe (Int, ByteString)
+{-# NOINLINE readInt #-}
+readInt = _readInt q
+  where q =
+# if WORD_SIZE_IN_BITS == 64
+            0xccccccccccccccc -- 0xccc_cccc_cccc_cccc
+# elif WORD_SIZE_IN_BITS == 32
+            0xccccccc         -- 0xccc_cccc
+# endif
+-- [Note @maxBound `quotRem` 10 == (q, 7)@]
+--
+-- It is not difficult to see that division of @2^n-1@ by @10@ produces a
+-- result that is a truncation of the infinite bitstring @1100_1100_...@.
+--
+-- When the input has 4n+3 bits, as is the case with 'maxBound' for all the
+-- Int* types, the quotient will have 4n bits with all the nibbles equal to
+-- @0xC@, and the remainder will be 7.
+--
+-- Would be nice to be able to use NumericUndescores some day, but that
+-- requires dropping support for GHC older than 8.6.1.
+
+-- | A variant of 'readInt' specialised to 'Int32'.
+readInt32 :: ByteString -> Maybe (Int32, ByteString)
+{-# NOINLINE readInt32 #-}
+readInt32 = _readInt 0xccccccc -- 0xccc_cccc
+
+-- | A variant of 'readInt' specialised to 'Int16'.
+readInt16 :: ByteString -> Maybe (Int16, ByteString)
+{-# NOINLINE readInt16 #-}
+readInt16 = _readInt 0xccc
+
+-- | A variant of 'readInt' specialised to 'Int8'.
+readInt8 :: ByteString -> Maybe (Int8, ByteString)
+{-# NOINLINE readInt8 #-}
+readInt8 = _readInt 0xc
+
+#else
+type Accum = Word64
+
+-- | A variant of 'readInt' specialised to 'Int64'.
+readInt64 :: ByteString -> Maybe (Int64, ByteString)
+{-# NOINLINE readInt64 #-}
+readInt64 = _readInt 0xccccccccccccccc -- 0xccc_cccc_cccc_cccc
+
+#endif
+
+----- Handle optional signs and dispatch to the word-reader
+
+data Sign = POS | NEG
+
+_readInt :: forall a. (Integral a, Bounded a)
+         => Accum -> ByteString  -> Maybe (a, ByteString)
+{-# INLINE _readInt #-}
+_readInt q bs =
+    params >>= \ (p, r, s) -> _readWord p q r s
+  where
+    -- See above [Note @maxBound `quotRem` 10 == (q, 7)@]
+    params = uncons bs >>= \ case
+        (w, s)
+            | fromDigit w <= 9 -> return (POS, 7, bs) -- leading digit
+            | w == 0x2d        -> return (NEG, 8, s)  -- minus sign
+            | w == 0x2b        -> return (POS, 7, s)  -- plus sign
+        _                      -> Nothing             -- not a number
+    fromDigit :: Word8 -> Word
+    fromDigit w = fromIntegral w - 0x30 -- i.e. w - '0'
+
+----- ByteString layer unsigned read
+
+-- See similar code in the "Lazy.ReadWord", where `r == 5` and
+-- only `q` is passed.  For cleaner specialisation we clone
+-- it here, rather that import the Word* code.  This also allows
+-- the Word* code to be slightly better/differently optimised.
+--
+_readWord :: forall a. (Integral a, Bounded a)
+          => Sign -> Accum -> Accum -> ByteString -> Maybe (a, ByteString)
+{-# INLINE _readWord #-}
+_readWord !sign !q !r =
+#ifdef BYTESTRING_STRICT
+    \s -> case _digits q r s 0 of
+        Result used acc
+            | used >= 1
+              -> convert acc $!
+                 let t = BU.unsafeDrop used s
```

Comment:
Is GHC clever enough to remove the allocation of the ByteString in the empty case? Otherwise replace with something like:

```haskell
if B.length s == used then empty else BU.unsafeDrop used s
```

Reply:
Yes, it is. This is visible in the generated Core above, near `case <=# t 0# of`.

Reply:
Yes, but if this is non-obvious, perhaps I should change it just to "unask the question"?

Reply:
I'll go ahead and do that. Same result, but more obvious.

Reply:
This change causes GHC to miss the relationship between the "isResultNull" condition and the length of the result in the non-empty case. Unfortunately, knowing that relationship causes at least GHC8.10.7 on x86-64 to generate worse assembly... It writes the condition as
```
        subq %r9,%rax
        testq %rax,%rax
        jg _caiK
```
and the extra `testq` nullifies any potential gain from avoiding a `subq` in the non-empty branch. I don't think there's a way to convince GHC's code generator to use `jns` here, so this change might represent a tiny performance win, too.